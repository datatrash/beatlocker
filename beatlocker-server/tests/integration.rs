use crate::test_utils::TestClient;
use axum::http::StatusCode;
use beatlocker_server::*;
use chrono::{DateTime, Utc};
use std::path::PathBuf;
use std::sync::Arc;

#[path = "test_utils/mod.rs"]
mod test_utils;

#[tokio::test]
async fn integration_test() -> AppResult<()> {
    enable_default_tracing();
    let options = ServerOptions {
        path: PathBuf::from("tests/data"),
        database: DatabaseOptions {
            path: Some(PathBuf::from(".")),
            in_memory: true,
        },
        now_provider: Arc::new(Box::new(|| {
            DateTime::parse_from_rfc3339("2020-02-02T00:00:00Z")
                .unwrap()
                .with_timezone(&Utc)
        })),
        ..Default::default()
    };
    let server = App::new(options).await?;
    let client = TestClient::new(server.app.clone());

    server.import_all_folders().await?;

    let res = client.get("/rest/ping?f=json").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!("ping.json", res.json::<serde_json::Value>().await);
    let res = client.get("/rest/ping").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_snapshot!("ping.xml", res.xml_string().await);

    let res = client.get("/rest/getLicense?f=json").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!("getLicense.json", res.json::<serde_json::Value>().await);
    let res = client.get("/rest/getLicense").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_snapshot!("getLicense.xml", res.xml_string().await);

    let res = client.get("/rest/getMusicFolders?f=json").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!(
        "getMusicFolders.json",
        res.json::<serde_json::Value>().await
    );
    let res = client.get("/rest/getMusicFolders").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_snapshot!("getMusicFolders.xml", res.xml_string().await);

    let motorway_ost_folder_uuid = "9600e0e0-145e-644d-d38f-e501a6252d79";
    let res = client
        .get(&format!(
            "/rest/getMusicDirectory?f=json&id={motorway_ost_folder_uuid}"
        ))
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!(
        "getMusicDirectory.json",
        res.json::<serde_json::Value>().await
    );
    let res = client
        .get(&format!(
            "/rest/getMusicDirectory?id={motorway_ost_folder_uuid}"
        ))
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_snapshot!("getMusicDirectory.xml", res.xml_string().await);

    let res = client
        .get("/rest/getIndexes?f=json&musicFolderId=00000000-0000-0000-0000-000000000000")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!("getIndexes.json", res.json::<serde_json::Value>().await);
    let res = client
        .get("/rest/getIndexes?musicFolderId=00000000-0000-0000-0000-000000000000")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_snapshot!("getIndexes.xml", res.xml_string().await);

    let res = client
        .get("/rest/getAlbumList?f=json&type=alphabeticalByName")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!("getAlbumList.json", res.json::<serde_json::Value>().await);
    let res = client
        .get("/rest/getAlbumList.view?type=alphabeticalByName")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_snapshot!("getAlbumList.xml", res.xml_string().await);

    let res = client
        .get("/rest/search3?f=json&query=\"\"&songCount=2")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!("search3.json", res.json::<serde_json::Value>().await);
    let res = client
        .get("/rest/search3?query=\"\"&songCount=2")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_snapshot!("search3.xml", res.xml_string().await);

    let res = client.get("/rest/getPlaylists?f=json").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!("getPlaylists.json", res.json::<serde_json::Value>().await);
    let res = client.get("/rest/getPlaylists").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_snapshot!("getPlaylists.xml", res.xml_string().await);

    let res = client
        .get(&format!(
            "/rest/getPlaylist?f=json&id={motorway_ost_folder_uuid}"
        ))
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!("getPlaylist.json", res.json::<serde_json::Value>().await);
    let res = client
        .get(&format!("/rest/getPlaylist?id={motorway_ost_folder_uuid}"))
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_snapshot!("getPlaylist.xml", res.xml_string().await);

    Ok(())
}
