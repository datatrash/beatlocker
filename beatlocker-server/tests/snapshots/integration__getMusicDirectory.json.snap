---
source: beatlocker-server/tests/integration.rs
expression: "res.json::<serde_json::Value>().await"
---
{
  "subsonic-response": {
    "directory": {
      "child": [
        {
          "created": "2020-02-02T00:00:00Z",
          "id": "694a6eee-8eac-2a60-af58-e96493edafcb",
          "isDir": true,
          "isVideo": false,
          "name": "MotorwayNested",
          "parent": "9600e0e0-145e-644d-d38f-e501a6252d79",
          "title": "MotorwayNested"
        },
        {
          "album": "Motorway (Original Motion Picture Soundtrack)",
          "albumId": "cbe80ce0-eb7e-37f6-a70a-87ab0dec3ac9",
          "artist": "Alex Gopher",
          "artistId": "c2042f2f-fbda-64e4-ff33-62bad6853d99",
          "bitRate": 160,
          "contentType": "audio/ogg",
          "created": "2020-02-02T00:00:00Z",
          "discNumber": 1,
          "duration": 6,
          "genre": "Unknown genre",
          "id": "7b2abff8-3571-99f6-b224-56250315eb14",
          "isDir": false,
          "isVideo": false,
          "parent": "9600e0e0-145e-644d-d38f-e501a6252d79",
          "size": 105424,
          "suffix": "ogg",
          "title": "Diamond Dealers",
          "track": 1,
          "year": 2021
        }
      ],
      "id": "9600e0e0-145e-644d-d38f-e501a6252d79",
      "name": "Motorway OST"
    },
    "serverVersion": "unknown",
    "status": "ok",
    "type": "beatlocker",
    "version": "1.16.1"
  }
}
